src/ray/raylet/scheduling/cluster_task_manager.cc

```diff
@@ -246,38 +246,58 @@ void ClusterTaskManager::DispatchScheduledTasksToWorkers(
-          // this task.
-          cluster_resource_scheduler_->ReleaseWorkerResources(allocated_instances);
-          ReleaseTaskArgs(task_id);
-          // It may be that no worker was available with the correct runtime env or
-          // correct job ID.  However, another task with a different env or job ID
-          // might have a worker available, so continue iterating through the queue.
-          work_it++;
-          // Keep track of runtime env that doesn't have workers available so we
-          // won't call PopWorker for subsequent tasks requiring the same runtime env.
-          blocked_runtime_env_to_skip.insert(runtime_env_worker_key);
-          continue;
-        }
-
-        RAY_LOG(DEBUG) << "Dispatching task " << task_id << " to worker "
-                       << worker->WorkerId();
-        auto reply = std::get<1>(*work_it);
-        auto callback = std::get<2>(*work_it);
-        Dispatch(worker, leased_workers_, allocated_instances, task, reply, callback);
-      }
-
-      if (!spec.GetDependencies().empty()) {
-        task_dependency_manager_.RemoveTaskDependencies(
-            task.GetTaskSpecification().TaskId());
+        work->waiting_worker_popped = true;
+        pending_workers_index_.emplace(task_id, work);
+        worker_pool_.PopWorker(
+            spec,
+            [this, allocated_instances, task_id, runtime_env_worker_key](
+                const std::shared_ptr<WorkerInterface> worker, Status status) {
+              auto it = pending_workers_index_.find(task_id);
+              RAY_CHECK(it != pending_workers_index_.end());
+              const auto &work = it->second;
+              const auto &reply = work->reply;
+              const auto &callback = work->callback;
+              const auto &canceled = work->canceled;
+              if (canceled || !worker) {
+                if (canceled) {
+                  RAY_LOG(DEBUG)
+                      << "Task " << task_id
+                      << " has been canceled when worker popped, maybe the task has "
+                         "been cancelled.";
+                } else {
+                  RAY_LOG(DEBUG)
+                      << "This node has available resources, but no worker processes "
+                         "to grant the lease "
+                      << task_id << ", status = " << status.CodeAsString();
+                }
+                // We've already acquired resources so we need to release them to avoid
+                // double-acquiring when the next invocation of this function tries to
+                // schedule this task.
+                cluster_resource_scheduler_->ReleaseWorkerResources(allocated_instances);
+                // No worker available, we won't be able to schedule any kind of task.
+                // Worker processes spin up pretty quickly, so it's not worth trying to
+                // spill this task.
+                ReleaseTaskArgs(task_id);
```

Comment:
It looks like
```cpp
      if (!spec.GetDependencies().empty()) {
        task_dependency_manager_.RemoveTaskDependencies(
            task.GetTaskSpecification().TaskId());
```
is always called even when worker == nullptr. Don't we need to also call it in this block?

Reply:
There is a if to process worker == nullptr?
https://github.com/ray-project/ray/blob/7378327c59af9bf48291a17eaee1b43c6ce832d0/src/ray/raylet/scheduling/cluster_task_manager.cc#L261

Reply:
Oh I meant this code block is called always in the original code (regardless of worker == nullptr or worker != nullptr)

Reply:
Great! My bug. Thank you. I fixed this.